backend/src/main/java/com/google/sps/notifs/EmailNotifier.java

```diff
@@ -0,0 +1,176 @@
+import java.io.InputStreamReader;
+import java.security.GeneralSecurityException;
+import java.util.List;
+import java.util.Properties;
+import javax.mail.Message.RecipientType;
+import javax.mail.MessagingException;
+import javax.mail.Session;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+import org.apache.commons.codec.binary.Base64;
+
+/** Class representing the notification system capable of sending mail to the user. */
+public class EmailNotifier {
+
+  private static final String APPLICATION_NAME = "Ad-lib";
+  private static final String APPLICATION_EMAIL = "Adlib-Step@gmail.com";
+
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+  private static final String TOKENS_DIRECTORY_PATH = "tokens";
+  /**
+   * Global instance of the scopes required by this quickstart. If modifying these scopes, delete
+   * your previously saved tokens/ folder.
+   */
+  private static final List<String> SCOPES = ImmutableList.of(GmailScopes.MAIL_GOOGLE_COM);
+
+  private static final String CREDENTIALS_FILE_PATH = "credentials.json";
+
+  /** The email to be notified */
+  private final String toEmail;
+
+  /** The notification recipient */
+  private final String recipientName;
+
+  /**
+   * @param recipientName Name of the recipient of the user
+   * @param toEmail The email addresse that this email is going to be sent to.
+   */
+  public EmailNotifier(String recipientName, String toEmail) {
+    this.recipientName = recipientName;
+    this.toEmail = toEmail;
+  }
+
+  /**
+   * Creates an authorized Credential object.
+   *
+   * @param HTTP_TRANSPORT The network HTTP Transport.
+   * @return An authorized Credential object.
+   * @throws IOException If the credentials.json file cannot be found.
+   */
+  private static Credential getCredentials(final NetHttpTransport HTTP_TRANSPORT)
+      throws IOException {
+    // Load client secrets.
+    InputStream in = new FileInputStream(CREDENTIALS_FILE_PATH);
+    GoogleClientSecrets clientSecrets =
+        GoogleClientSecrets.load(JSON_FACTORY, new InputStreamReader(in));
+
+    // Build flow and trigger user authorization request.
+    GoogleAuthorizationCodeFlow flow =
+        new GoogleAuthorizationCodeFlow.Builder(HTTP_TRANSPORT, JSON_FACTORY, clientSecrets, SCOPES)
+            .setDataStoreFactory(new FileDataStoreFactory(new File(TOKENS_DIRECTORY_PATH)))
+            .setAccessType("offline")
+            .build();
+    LocalServerReceiver receiver = new LocalServerReceiver.Builder().setPort(8000).build();
+    return new AuthorizationCodeInstalledApp(flow, receiver).authorize("user");
+  }
+
+  /**
+   * Create a MimeMessage using the parameters provided.
+   *
+   * @param toEmail email address of the receiver
+   * @param subject subject of the email
+   * @param bodyText body text of the email
+   * @return the MimeMessage to be used to send email
+   * @throws MessagingException for other failures
```

Comment:
Can you expand on the throws doc here. What failures are possible?

Reply:
There are a lot of other failures , I don't know how to encapsulate them as the are all referred to here : https://docs.oracle.com/javaee/7/api/javax/mail/class-use/MessagingException.html

Reply:
Which calls in this method can throw this exception? Under what circumstances will they throw?